### PR TITLE
Accessibility improvements

### PR DESCRIPTION
**Pull request by sha-265**
# Description

Relates to issue: https://github.com/ZeusLN/zeus/issues/1442

Added accessibility improvements as requested by the folks at the Bitcoin Design Community that have detailed in the above issue. Should be tested on iOS also.

How do I get the bounty? 🙂

This pull request is categorized as a:

- [ ] New feature
- [ ] Bug fix
- [ ] Code refactor
- [ ] Configuration change
- [ ] Locales update
- [ ] Quality assurance 
- [x] Other

## Checklist
- [ ] I’ve run `yarn run tsc` and made sure my code compiles correctly
- [x] I’ve run `yarn run lint` and made sure my code didn’t contain any problematic patterns
- [x] I’ve run `yarn run prettier` and made sure my code is formatted correctly
- [x] I’ve run `yarn run test` and made sure all of the tests pass

## Testing

If you modified or added a utility file, did you add new unit tests?

- [ ] No, I’m a fool
- [ ] Yes
- [ ] N/A

I have tested this PR on the following platforms (please specify OS version and phone model/VM):

- [x] Android
- [ ] iOS

Tested on Android emulator (Pixel 3a with android 12). Probably should be tested on iOS. Maybe @mouxdesign would want to test it more.

I have tested this PR with the following types of nodes (please specify node version and API version where appropriate):

- [ ] LND (REST)
- [ ] LND (Lightning Node Connect)
- [x] Core Lightning (c-lightning-REST)
- [ ] Core Lightning (Spark)
- [ ] Eclair
- [ ] LndHub

### Locales
- [ ] I’ve added new locale text that requires translations
- [ ] I’m aware that new translations should be made on the Zeus [Transfix page](https://app.transifex.com/ZeusLN/zeus/) and not directly to this repo

### Third Party Dependencies and Packages

- [ ] Contributors will need to run `yarn` after this PR is merged in
- [ ] 3rd party dependencies have been modified:
    * verify that `package.json` and `yarn.lock` have been properly updated
    * verify that dependencies are installed for both iOS and Android platforms

### Other:

- [ ] Changes were made that require an update to the README
- [ ] Changes were made that require an update to onboarding

### CODE REVIEW

**Comment by kaloudis** [comment]
@sha-265 nice work!

We'd like to have the Bitcoin Design Community review this (ideally @mouxdesign), then once it's merged we'll pay you out.

**Comment by sha-265** [comment]
IIUC, @GBKS made the initial test, so maybe he is relevant too.

**Comment by GBKS** [comment]
Thanks a ton for working on these fixes. I'd love to test, but am not able to run the application locally. My technical expertise is miserably failing at some Ruby-related issues related to paths that I can't make sense of. Typical result in these situations is that I bang my head against it for several hours with lots of Googling and AI consultation and still can't resolve it. Are there other options for testing?

**Comment by kaloudis** [comment]
@GBKS we could potentially make a build for you on your platform of choice

**Comment by sha-265** [comment]
Android APKs are available [here](https://github.com/sha-265/zeus/releases/tag/v0.7.7-accessibitilty-improvments), if it helps somehow.

**Comment by GBKS** [comment]
@sha-265 I tried to install the APK, but Android tells me the package is invalid. 🤷‍♂️

**Comment by kaloudis** [comment]
@sha-265 @GBKS the APKs are unsigned so Android will not allow you to install them

https://developer.android.com/studio/publish/app-signing

**Comment by sha-265** [comment]
Hi @GBKS,

I uploaded signed universal APK [here](https://github.com/sha-265/zeus/releases/tag/v0.7.7-accessibitilty-improvments). Please try it.

**Comment by sha-265** [comment]
@GBKS any updates?

**Comment by GBKS** [comment]
Sorry for the delay, I am still catching up with projects. Below are my first test results. 

It is a bit difficult for me to match all the code changes to what I see in the UI, so I tried to simply fully test the screens that I identified changes on. If I left a note on something not covered by this PR, feel free to ignore (or open a separate issue, or address, as you like).

Please also let me know what I missed (just ran into another tech issue just now and had to stop testing). I think the only screens I still need to test are the Nodes and Currency settings screens, but I hope this is already helpful.

I hope the notes are self-explanatory. A checkmark is for an element that I tested and found everything to be good. Otherwise I used a text note.

Zeus accessibility review 230905

One thing across many screens is that the amount and unit are separately selectable. They are probably separate elements to achieve the nice visual treatment. Unifying those for screen readers would be nice, not sure if that's worth being its own component (maybe it already is)?

Also cool if you'd like to merge these changes and open another PR for more accessibility tweaks. This PR seems like a good step forward and might be helpful to users that rely on screen readers, etc.

Anyhow, hope this helps, thanks for your patience, and please let me know what I missed.

**Comment by GBKS** [comment]
Just a quick extra note on my test process:

- I use a Galaxy A32 5G with TalkBack enabled.
- I try to test without looking at the screen as much as possible (but have to look at it regularly for taking notes and screenshots).
- I swipe left/right to select different elements. The device announces what is selected (label, element type, etc).
- I double-tap to activate the selected element, expecting to be informed about what happened.

Pretty straightforward overall. It's a really interesting experience to try to navigate an app without looking, to see if you can piece the bits from the screenreader together. We're biased of course, because we have seen the screens before, so we have mental modals that a blind user won't have.

**Comment by sha-265** [comment]
> Also cool if you'd like to merge these changes and open another PR for more accessibility tweaks. This PR seems like a good step forward and might be helpful to users that rely on screen readers, etc.

@kaloudis would you like to merge this changes now? Or should I add the other changes before to this PR?

**Comment by sha-265** [comment]
@GBKS I fixed the issues you mentioned. If you want to check again I will try to create new APK.

**Comment by sha-265** [comment]
Any updates?

**Comment by GBKS** [comment]
Thank you for continuing with these improvements. I tested and it looks great. There are two areas that could be improved, as hopefully explained well on the images below. Since visually impaired users can't see what happens, we just have to be clear in the labels, announce changes, and labels need to change when the button state changes.

image

I'm excited to get this wrapped up and out there. 🪩

**Comment by sha-265** [comment]
@GBKS, 

1. It can read before the the number AFAIK (not easily at least). I can only make it read "Total balance" and when continue navigating, it will read the number separately. Is that OK?
2. Making these swipeable rows accessibility labels dynamically change is to complex for me. I'm not an expert.
3. For me the channels header is not a toggle. What version are you using?

**Comment by kaloudis** [comment]
@sha-265 the channel toggle will only appear if you're using LND

**Comment by GBKS** [comment]
Totally good if you want to merge this PR, and keep the other tweaks as separate tasks for someone who has dealt with those specific things before. As a self-declared non-expert, you've been doing a fantastic job getting things done, big thanks for that.

Making the screen reader read out changes seems fairly straightforward from some quick research (famous last words). There is an "[aria-live](https://reactnative.dev/docs/accessibility#aria-live-android)" tag you can add to elements that tells the screen reader to pay attention to it and announce any changes to that element. So if you add that tag, and change the accessibility label based on open/closed state, it _should_ do the trick. I'd also like to know how to solve this because I see this problem in a lot of apps.

**Comment by sha-265** [comment]
> @sha-265 the channel toggle will only appear if you're using LND

@kaloudis, that explains it. I'm using CLN and I don't have LND test env.



> Making the screen reader read out changes seems fairly straightforward from some quick research (famous last words). There is an "[aria-live](https://reactnative.dev/docs/accessibility#aria-live-android)" tag you can add to elements that tells the screen reader to pay attention to it and announce any changes to that element. So if you add that tag, and change the accessibility label based on open/closed state, it _should_ do the trick. I'd also like to know how to solve this because I see this problem in a lot of apps.

@GBKS, I tried to use `aria-live` but no success. I think `aria-live` is working when the element's content itself is changing. In our case the content is not changing, but there is only an animation of the element.

I got into a state that I can update the accessibility state of the element, but I think it's not re-rendering the element with the new state. AFAIU we can use [direct manipulation](https://reactnative.dev/docs/0.70/direct-manipulation), or re-rendering the element somehow else.

I tried to do something like that:
```
 onSwipeableOpen={() => { children.props.accessibilityState = {{expanded: true}} }
 onSwipeableClose={() => { children.props.accessibilityState = {{expanded: false}} }
```

But as I said the state is changing but it's not being rendered. Maybe @kaloudis can help with this?

**Comment by GBKS** [comment]
Sorry, I don't have any more tips. Seems like something that requires a bit of experimentation and testing hands-on with different techniques. What do you think @kaloudis?

**Comment by kaloudis** [comment]
@sha-265 a better approach for `LayerBalances` may be to try to track the state of them (open/closed) in the component state. Would have to take a deeper look.

Also, lmk if you want an LND testnet node to work with. Would gladly give you access for this effort.

**Comment by sha-265** [comment]
> Also, lmk if you want an LND testnet node to work with. Would gladly give you access for this effort.

Yes, please.

**Comment by kaloudis** [comment]
> > Also, lmk if you want an LND testnet node to work with. Would gladly give you access for this effort.
> 
> Yes, please.

Sent you the credentials

**Comment by sha-265** [comment]
I fixed the channels header toggle accessibility problem using `accessibilityLiveRegion` (set to `polite`), but this property is relevant only to Android. There are maybe other solutions for iOS like [this one](https://stackoverflow.com/questions/63812424/react-native-accessibilityliveregion-equivalent-in-ios), but anyway I can't test it on iOS, so I didn't try it.

File: locales/en.json [hunk 1]
```diff
@@ -125,12 +127,17 @@
...
+    "components.QRCodeScanner.chooseFromGallery": "Choose from gallery",
+    "components.QRCodeScanner.flashOn": "Flash on",
+    "components.QRCodeScanner.flashOff": "Flash off",
     "models.Channel.unknownId": "Unknown Channel ID",
     "models.Invoice.noMemo": "No memo",
     "models.Invoice.seconds": "seconds",
     "models.Invoice.never": "Never",
     "models.Payment.forwarded": "Forwarded",
     "models.Payment.fee": "Fee",
+    "views.index.accessibilityLabel.On-chain": "View on-chain options",
```
[the remark this answers is not in the document]
Comment (sha-265) [reply locales/en.json:139]: it is being use [here](https://github.com/sha-265/zeus/blob/4fab7251307df4608f7e48401a78fe2d36cd478c/components/LayerBalances/index.tsx#L49)

**Pull request merged**